src/PureScript/Backend/Chez/Convert.purs

```diff
@@ -10,31 +10,50 @@ import Data.Maybe (Maybe(..))
 import Data.Newtype (unwrap, wrap)
 import Data.Set as Set
 import Data.String.CodeUnits as CodeUnits
+import Data.Tuple as Tuple
 import Data.Tuple (Tuple(..), uncurry)
+import Partial.Unsafe (unsafeCrashWith)
 import PureScript.Backend.Chez.Syntax (ChezDefinition(..), ChezExport(..), ChezExpr, ChezImport(..), ChezImportSet(..), ChezLibrary)
 import PureScript.Backend.Chez.Syntax as S
 import PureScript.Backend.Optimizer.Convert (BackendModule, BackendBindingGroup)
-import PureScript.Backend.Optimizer.CoreFn (Ident(..), Literal(..), ModuleName(..), Qualified(..))
+import PureScript.Backend.Optimizer.CoreFn (Ident(..), Literal(..), ModuleName(..))
 import PureScript.Backend.Optimizer.Semantics (NeutralExpr(..))
-import PureScript.Backend.Optimizer.Syntax (BackendOperator(..), BackendOperator1(..), BackendOperator2(..), BackendOperatorNum(..), BackendOperatorOrd(..), BackendSyntax(..), Pair(..))
+import PureScript.Backend.Optimizer.Syntax (BackendAccessor(..), BackendOperator(..), BackendOperator1(..), BackendOperator2(..), BackendOperatorNum(..), BackendOperatorOrd(..), BackendSyntax(..), Pair(..))
 import Safe.Coerce (coerce)
 
 type CodegenEnv =
   { currentModule :: ModuleName
   }
 
+type DefinitionTranslation =
+  { pursIdent :: Ident
+  , chezDefinitions :: Array ChezDefinition
+  }
+
 codegenModule :: BackendModule -> ChezLibrary
 codegenModule { name, bindings, exports, imports, foreign: foreign_ } =
   let
     codegenEnv :: CodegenEnv
     codegenEnv = { currentModule: name }
 
-    exports' :: Array ChezExport
-    exports' = map ExportIdentifier $ coerce $ Array.fromFoldable exports
+    translations :: Array DefinitionTranslation
+    translations = Array.concatMap (codegenTopLevelBindingGroup codegenEnv) bindings
 
     definitions :: Array ChezDefinition
-    definitions = Array.concat $
-      codegenTopLevelBindingGroup codegenEnv <$> bindings
+    definitions = Array.concatMap _.chezDefinitions translations
+
+    exports' :: Array ChezExport
+    exports' = map ExportIdentifier
+      $ Array.sort
+      $ Array.concatMap S.definitionIdentifiers definitions'
+          <> map coerce (Array.fromFoldable foreign_)
+      where
+      translations' :: Array DefinitionTranslation
+      translations' =
+        Array.filter (\t -> Set.member t.pursIdent exports) translations
```

Comment:
Could you clarify why you didn't add `exports` to the `codegenEnv`, and then determine whether a definition should be exported in `codegenToplevelBinding`?

I think that would clean up this code.